### PR TITLE
Add public API to pin and unpin presets to the sample cache

### PR DESCRIPTION
**Pull request by mawe42**
Following the discussion about an API to pin and unpin preset samples in the sample cache here:
https://lists.nongnu.org/archive/html/fluid-dev/2020-10/msg00016.html

This is a request for comment pull-request, the code has not been tested yet. I only checked if it would compile and am presenting it here to get feedback on the approach.

Short explanation of the change:

Only the default loader currently supports dynamic sample loading, so I thought it might be a good idea to keep the changes for this feature mostly contained in the default loader as well. I've added two new preset notify flags (`FLUID_PRESET_PIN` and `FLUID_PRESET_UNPIN`) that are handled by the `preset->notify` callback and trigger the loading and possibly unloading of the samples.

The advantage of this approach is that changes to existing datastructures are contained to the internal default loader. The downside is that calls to the new `fluid_synth_pin_preset` and `fluid_synth_unpin_presets` can not return any information about if the pinning or unpinning has actually succeeded.

Looking forward to your thoughts!

### CODE REVIEW

**Comment by jjceresa** [comment]
This new functions integrates naturally in the existing `dynamic sample loading` logic. This look good.

>The downside is that calls to the new fluid_synth_pin_preset and fluid_synth_unpin_presets can not return any information about if the pinning or unpinning has actually succeeded.

This `absence of information` is not particular to the 2 new pin/unpin functions. This `absence` was already the case within `fluid_channel_set_preset()`. This is due to the fact that we use the macro  `fluid_preset_notify()` that ignores the code returned by the callback `dynamic_samples_preset_notify()`. Similarly `dynamic_samples_preset_notify()` function  ignores the code returned by `load_preset_samples()` and `unload_preset_samples()` .
`load_preset_samples()` ignores the code returned by `fluid_defsfont_load_sampledata()`.
`unload_preset_samples()`  cannot return a code when calling `unload_sample()` which return void.

Perhaps this could be changed but it doesn't worth (imo) because the user will got an audible feedback that something is wrong in case of failure. Also we have the DEBUG log that should help, but it seems incomplete:
`unload_samples()` will print message a FLUID_DBG message but `fluid_defsfont_load_sampledata()` doesn't, so we need to add a FLUID_DBG message in `fluid_defsfont_load_sampledata()`.

**Comment by mawe42** [comment]
> This `absence of information` is not particular to the 2 new pin/unpin functions. This `absence` was already the case within `fluid_channel_set_preset()`. [...]
> 
> Perhaps this could be changed but it doesn't worth (imo) because the user will got an audible feedback that something is wrong in case of failure.

Good point! And I agree, it could be cleaned up, but it's not really that important at the moment.

> Also we have the DEBUG log that should help, but it seems incomplete:
> `unload_samples()` will print message a FLUID_DBG message but `fluid_defsfont_load_sampledata()` doesn't, so we need to add a FLUID_DBG message in `fluid_defsfont_load_sampledata()`.

Also a good point. I will add that to this PR if we decide to go ahead with this.

Thanks for your review!

**Comment by mawe42** [comment]
I've added some tests for the pinning and unpinning. Seems to work fine, although I haven't tested in an actual application yet.

To be able to test this properly, I added a new helper function to count the number of entries in the samplecache (
11d821b). I would really like to get your opinion on that. Do you think test helper functions like that are acceptable? I'm in two minds about it... code that only exists for tests should really be in the test codebase. But having a reliable way to see whats happening in the samplecache makes tests so much easier, so I thought it would be a good addition.

**Comment by mawe42** [comment]
That is strange, the WindowsCI debug_prof build fails on one of the pinning tests:
https://dev.azure.com/tommbrt/tommbrt/_build/results?buildId=2499&view=logs&j=fac68015-cd32-5143-94f9-ac37c5d8f8b3&t=32193c98-3008-58b0-a8ec-6a41f54abb0c&l=1259

I have no idea what could cause this... do you?

**Review by derselbst — COMMENTED** [review]
First of all, the code looks really nice and clean. Good job!

> > The downside is that calls to the new fluid_synth_pin_preset and fluid_synth_unpin_presets can not return any information about if the pinning or unpinning has actually succeeded.
> 
> This `absence of information` is not particular to the 2 new pin/unpin functions. This `absence` was already the case within `fluid_channel_set_preset()`. This is due to the fact that we use the macro `fluid_preset_notify()` that ignores the code returned by the callback `dynamic_samples_preset_notify()`. Similarly `dynamic_samples_preset_notify()` function ignores the code returned by `load_preset_samples()` and `unload_preset_samples()` .
> `load_preset_samples()` ignores the code returned by `fluid_defsfont_load_sampledata()`.
> `unload_preset_samples()` cannot return a code when calling `unload_sample()` which return void.

We must make sure that `fluid_synth_pin_preset()` returns the correct result. Relying on audible feedback is not sufficient. And apart from what JJC already said, I currently see no bigger problem, why returning this information should not be possible or hard to implement.

> I added a new helper function to count the number of entries in the samplecache. Do you think test helper functions like that are acceptable?

I think it's fine.

So, all I can say is, give it test :) And don't worry too much about the failing test, I can assist with that upcoming weekend.

**Comment by mawe42** [comment]
> First of all, the code looks really nice and clean. Good job!

Thanks :-)

> We must make sure that `fluid_synth_pin_preset()` returns the correct result. Relying on audible feedback is not sufficient.

Ok, I'll add all the missing return paths.

> So, all I can say is, give it test :)

I will make the return value changes first, adjust the tests and then try to integrate it into my application.

> And don't worry too much about the failing test, I can assist with that upcoming weekend.

Luckily I can reproduce it on my machine. With build type Debug, the test fails. Really strange!

**Comment by mawe42** [comment]
Fixed the test failure on Debug builds, it was a missing initialization of the new `pinned` struct member.

I must say, I *really* like our CI infrastructure! :-)

File: test/test_preset_pinning.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    TEST_ASSERT(fluid_synth_pin_preset(synth, id, 0, 42) == FLUID_OK);
+    TEST_ASSERT(count_loaded_samples(synth, id) == 4);
+    TEST_ASSERT(fluid_samplecache_count_entries() == 4);
+
+    TEST_ASSERT(fluid_synth_sfunload(synth, id, 0) == FLUID_OK);
+    TEST_ASSERT(fluid_samplecache_count_entries() == 0);
+
+
+    /* Test that deleting the synth with a pinned preset also leaves no
+     * samples in cache */
```
[the remark this answers is not in the document]
Comment (mawe42) [reply test/test_preset_pinning.c:119]: You mean you would expect it a few lines further down? Those heading were meant as a description of the following blocks of code. And here the following block test exactly what is says in the comment. Or am I missing something obvious?

**Comment by jjceresa** [comment]
>With build type Debug, the test fails. Really strange!

I am not sure what you mean by "test fails", but here `test_preset_pinning.exe` (Release or Debug) run without assertion.
here the output:

D:\WinProjet\Git\GitHub\fluidsynth-VS10\test\Release>test_preset_pinning.exe
fluidsynth: debug: Testing audio device: Realtek HD Audio output
fluidsynth: debug: Testing audio device: SB Live! Audio [CE00]
fluidsynth: debug: Testing audio device: SB PCI
fluidsynth: debug: Testing audio device: Audio Bluetooth
fluidsynth: warning: Sample 'SineWave': ROM sample ignored
fluidsynth: Loaded samples on sfont 1: 123

fluidsynth: Loaded samples on sfont 1: 123

fluidsynth: Loaded samples on sfont 1: 123

fluidsynth: error: There is no preset with bank number 42 and preset num
fluidsynth: Loaded samples on sfont 1: 123

fluidsynth: error: There is no preset with bank number 42 and preset num
fluidsynth: Loaded samples on sfont 1: 123

fluidsynth: warning: Sample 'SineWave': ROM sample ignored
fluidsynth: Loaded samples on sfont 1: 0

fluidsynth: debug: Pinning preset 'Lead Synth 2'
fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: debug: Unpinning preset 'Lead Synth 2'
fluidsynth: debug: Unloading sample 'Saw_a4'
fluidsynth: debug: Unloading sample 'Saw_a3'
fluidsynth: debug: Unloading sample 'Saw_a2'
fluidsynth: debug: Unloading sample 'Saw_a1'
fluidsynth: Loaded samples on sfont 1: 0

fluidsynth: debug: Pinning preset 'Lead Synth 2'
fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: debug: Selected preset 'Lead Synth 2' on channel 0
fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: debug: Selected preset 'Aluminium Plate' on channel 1
fluidsynth: Loaded samples on sfont 1: 5

fluidsynth: debug: Deselected preset 'Lead Synth 2' from channel 0
fluidsynth: Loaded samples on sfont 1: 5

fluidsynth: debug: Deselected preset 'Aluminium Plate' from channel 1
fluidsynth: debug: Unloading sample 'Noise4'
fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: debug: Unpinning preset 'Lead Synth 2'
fluidsynth: debug: Unloading sample 'Saw_a4'
fluidsynth: debug: Unloading sample 'Saw_a3'
fluidsynth: debug: Unloading sample 'Saw_a2'
fluidsynth: debug: Unloading sample 'Saw_a1'
fluidsynth: Loaded samples on sfont 1: 0

fluidsynth: debug: Selected preset 'Lead Synth 2' on channel 0
fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: debug: Deselected preset 'Lead Synth 2' from channel 0
fluidsynth: debug: Unloading sample 'Saw_a4'
fluidsynth: debug: Unloading sample 'Saw_a3'
fluidsynth: debug: Unloading sample 'Saw_a2'
fluidsynth: debug: Unloading sample 'Saw_a1'
fluidsynth: Loaded samples on sfont 1: 0

fluidsynth: debug: Pinning preset 'Lead Synth 2'
fluidsynth: Loaded samples on sfont 1: 4

fluidsynth: debug: Unpinning preset 'Lead Synth 2'
fluidsynth: debug: Unloading sample 'Saw_a4'
fluidsynth: debug: Unloading sample 'Saw_a3'
fluidsynth: debug: Unloading sample 'Saw_a2'
fluidsynth: debug: Unloading sample 'Saw_a1'
fluidsynth: debug: Unloaded SoundFont
fluidsynth: warning: Sample 'SineWave': ROM sample ignored
fluidsynth: debug: Pinning preset 'Lead Synth 2'
fluidsynth: Loaded samples on sfont 2: 4

fluidsynth: debug: Unpinning preset 'Lead Synth 2'
fluidsynth: debug: Unloading sample 'Saw_a4'
fluidsynth: debug: Unloading sample 'Saw_a3'
fluidsynth: debug: Unloading sample 'Saw_a2'
fluidsynth: debug: Unloading sample 'Saw_a1'

**Comment by mawe42** [comment]
> I am not sure what you mean by "test fails", but here `test_preset_pinning.exe` (Release or Debug) run without assertion.

Thanks for testing it! The error I meant was this:
https://dev.azure.com/tommbrt/tommbrt/_build/results?buildId=2499&view=logs&j=fac68015-cd32-5143-94f9-ac37c5d8f8b3&t=32193c98-3008-58b0-a8ec-6a41f54abb0c&l=1259

But it's fixed now, missing initialization of `defpreset->pinned` in `new_fluid_defpreset`.

**Comment by jjceresa** [comment]
From your recent changes test_preset_pinning.c fail to compile, so I fix that in next commit.

**Comment by mawe42** [comment]
> From your recent changes test_preset_pinning.c fail to compile, so I fix that in next commit.

Great, thanks! I sometimes forget that we need to be C89 compatible here.

File: test/test_preset_pinning.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (derselbst) [reply test/test_preset_pinning.c:155]: Initially, my feeling said the comment should be around the `delete_fluid_synth()` call. But looking at it again, it's good.

**Comment by derselbst** [comment]
One open point is left: The DLS preset should return FLUID_FAILED when notified via PIN / UNPIN. I could look into this on the weekend.

**Comment by mawe42** [comment]
> One open point is left: The DLS preset should return FLUID_FAILED when notified via PIN / UNPIN. I could look into this on the weekend.

Ah, yes. I could add a new macro that checks if `preset->notify` is set and returns false if it isn't. Assuming my assumption is correct that DLS presets do not use the preset notification system. Or insert that code directly into the pin and unpin functions, without a macro...

**Comment by mawe42** [comment]
Hm, but one could also argue that as DLS presets are never loaded with dynamic-sample-loading, it's very similar to the case then calling `fluid_synth_pin_preset` with dynamic-sample-loading disabled. You said that with d-s-l disabled, calls to the pinning functions should return FLUID_OK, as the samples are already in memory.

The same logic applies to DLS samples with d-s-l enabled: all the samples are already in memory, so it would be consistent to also return FLUID_OK, as currently implemented.

**Comment by derselbst** [comment]
> Hm, but one could also argue that as DLS presets are never loaded with dynamic-sample-loading, it's very similar to the case when calling fluid_synth_pin_preset with dynamic-sample-loading disabled

Exactly my thought, after having written my comment above... I'm just 3 hours too late. Ok, let's keep the FLUID_OK logic, I'll amend the API comment.

**Comment by mawe42** [comment]
Thanks for rewriting the comments and adding the devdoc entries!

I've tested the feature in my application now: it seems to work as intended and helps me to remove the ugly hack I had used previously. So from my perspective, this is finished.

The question remains if it is a feature we actually want to include, adding yet more public API functions. So far I seem to be the only one with a use-case for it...

**Review by derselbst — APPROVED** [review]
> The question remains if it is a feature we actually want to include, adding yet more public API functions.

Yes, we want. It's a feature that didn't exist before and that nicely integrates into the existing architecture. (Unlike PR 672, for which I'm uncertain if it will be already included in 2.2.0, as we need to talk about it again).

So, allow me to squash it once the build is green.

**Comment by mawe42** [comment]
> PR 672, for which I'm uncertain if it will be already included in 2.2.0, as we need to talk about it again

Ah, yes. Reading the thread in #669 again, you suggested that I start a discussion about that on the mailing list. I will try to get back into the topic and then make a post so we can discuss it further, ok?

> So, allow me to squash it once the build is green.

Sounds good to me, thanks!

**Comment by derselbst** [comment]
> I will try to get back into the topic and then make a post so we can discuss it further, ok?

Perfect.

**Pull request merged**